lib/posix-process/wait.c

```diff
@@ -1,78 +1,262 @@
+
+	UK_ASSERT(pid > 0 || pid == -1);
+	UK_ASSERT(si);
+
+	this_thread = uk_pthread_current();
+	UK_ASSERT(this_thread);
+
+	this_process = this_thread->process;
+	UK_ASSERT(this_process);
+
+	UK_ASSERT(!uk_list_empty(&this_process->children));
+
+	this_thread->state = POSIX_THREAD_BLOCKED_WAIT;
 
-/* FIXME: Provide with sys/wait.h */
-struct rusage;
-typedef unsigned int idtype_t;
-typedef unsigned int id_t;
-typedef unsigned int siginfo_t;
+	if (pid > 0) {
+		child = pid2pprocess(pid);
+		/* Child does not exist or not our child. This also
+		 * covers the case where the process chose to ignore
+		 * SIGCHLD or set SA_NOCLDWAIT and therefore _exit()
+		 * terminated it without further ado.
+		 */
+		if (unlikely(!child || child->parent != this_thread->process)) {
+			ret = -ECHILD;
+			goto out;
+		}
 
+		/* If the process exited already, reap and return the pid */
+		if (child->state != POSIX_PROCESS_RUNNING) {
+			pprocess_reap(child, si);
+			goto out;
+		}
+
+		this_thread->wait_pid = pid;
+	} else { /* any */
+		this_thread->wait_pid = UK_PID_WAIT_ANY;
+	}
+
+	for (;;) {
+		/* Find process with changed state. No change means that
+		 * the state change happened to a different child, so we
+		 * loop back to the semaphore.
+		 */
+		if (pid > 0)
+			waitable = pprocess_state_changed(child, opt);
+		else /* any */
+			waitable = pprocess_child_state_changed(opt);
+
+		if (waitable) {
+			ret = waitable->pid;
+			if (!(opt & WNOWAIT))
+				pprocess_reap(waitable, si);
+			goto out;
+		}
+
+		if (opt & WNOHANG) {
+			ret = 0;
+			break;
+		}
+
+		/* Wait for a state change on one of our children */
+		uk_semaphore_down(&this_thread->process->wait_semaphore);
+	}
+
+out:
+	this_thread->state = POSIX_THREAD_RUNNING;
+	return ret;
+}
+
+pid_t uk_posix_process_wait(void)
+{
+	siginfo_t si;
+	int ret;
+
+	if (uk_list_empty(&uk_pprocess_current()->children))
+		return -ECHILD;
+
+	ret = pprocess_wait(-1, &si, WEXITED);
+
+	return ret > 0 ? si.si_pid : ret;
+}
+
+UK_SYSCALL_R_DEFINE(pid_t, wait4, pid_t, pid,
+		    int *, wstatus, int, options,
+		    struct rusage * __unused, rusage)
+{
+	siginfo_t si;
+	int ret;
+
+	if (options)
+		if (unlikely(!(options & (WNOHANG | WUNTRACED | WCONTINUED))))
+			return -EINVAL;
+
+	if (uk_list_empty(&uk_pprocess_current()->children))
+		return -ECHILD;
+
+	if (pid < -1 || pid == 0) /* process groups */
+		return -ENOTSUP;
+
+	ret = pprocess_wait(pid, &si, options | WEXITED);
+
+	if (ret > 0 && wstatus) {
+		if (si.si_status == CLD_KILLED)
+			*wstatus = si.si_code;
+		else /* CLD_EXITED */
+			*wstatus = si.si_code << 8;
+	}
+
+	return ret > 0 ? si.si_pid : ret;
+}
+
+UK_LLSYSCALL_R_DEFINE(int, waitid, idtype_t, idtype, id_t, id,
+		      siginfo_t *, infop, int, options,
+		      struct rusage * __unused, rusage)
+{
+	int ret;
+
+	if (options)
+		if (unlikely(!(options & (WEXITED | WSTOPPED | WCONTINUED |
+					  WNOHANG | WNOWAIT))))
+			return -EINVAL;
+
+	if (uk_list_empty(&uk_pprocess_current()->children))
```

Comment:
Is this common or would it be worth an `unlikely`? Ditto `wait4`/`uk_posix_process_wait`

Reply:
I would say, neither likely or unlikely. It depends on the application, so I refrained from making assumptions.